Cython/Utility/ModuleSetupCode.c

```diff
@@ -636,6 +636,32 @@ static CYTHON_INLINE void * PyThread_tss_get(Py_tss_t *key) {
 // PyThread_ReInitTLS() is a no-op
 #endif /* TSS (Thread Specific Storage) API */
 
+
+#if PY_MAJOR_VERSION < 3
+#if (CYTHON_COMPILING_IN_PYPY && PYPY_VERSION_NUM < 0x07030600)
+    #if defined(__cplusplus) && __cplusplus >= 201402L
+        [[deprecated("`with nogil:` inside a nogil function will not release the GIL in PyPy2")]]
+    #elif defined(__GNUC__) || defined(__clang__)
+        __attribute__ ((__deprecated__("`with nogil:` inside a nogil function will not release the GIL in PyPy2")))
+    #elif defined(_MSC_VER)
+        __declspec(deprecated("`with nogil:` inside a nogil function will not release the GIL in PyPy2"))
+    #endif
```

Comment:
I'd give them at least a version. It seems confusing and backwards to get a *deprecation* warning for something that is essentially a bug which was already resolved in a future version.
```suggestion
    #if defined(__cplusplus) && __cplusplus >= 201402L
        [[deprecated("`with nogil:` inside a nogil function will not release the GIL in PyPy2 < 7.3.6")]]
    #elif defined(__GNUC__) || defined(__clang__)
        __attribute__ ((__deprecated__("`with nogil:` inside a nogil function will not release the GIL in PyPy2 < 7.3.6")))
    #elif defined(_MSC_VER)
        __declspec(deprecated("`with nogil:` inside a nogil function will not release the GIL in PyPy2 < 7.3.6"))
    #endif
```

Reply:
OTOH, if the comment is correct, and we're actually not releasing the GIL when users ask for it, then this seems a regression, right?

Reply:
Sort of. The problem is that (at least on CPython and probably PyPy) it's a hard crash to try to release the GIL when we don't hold it.

It's currently allowed in Cython 0.29 (but will crash if the GIL isn't held) but blocked in 3.0.alpha. So it isn't a regression compared to the current master.

It also only applies to a `with nogil` block in a `nogil` function. Most of the time it knows it has the GIL so skips the check